src/charm.py

```diff
@@ -32,140 +31,144 @@
-        self.framework.observe(self.on.prometheus_pebble_ready, self._configure)
-        self.framework.observe(self.on.config_changed, self._configure)
-        self.framework.observe(self.on.upgrade_charm, self._configure)
-        self.framework.observe(self.ingress.on.ready_for_unit, self._configure)
-        self.framework.observe(self.on.receive_remote_write_relation_created, self._configure)
-        self.framework.observe(self.on.receive_remote_write_relation_changed, self._configure)
-        self.framework.observe(self.on.receive_remote_write_relation_broken, self._configure)
-        self.framework.observe(self.metrics_consumer.on.targets_changed, self._configure)
-        self.framework.observe(self.alertmanager_consumer.on.cluster_changed, self._configure)
-        self.framework.observe(
-            self.on.validate_configuration_action, self._on_validate_configuration
+        self.alertmanager_consumer = AlertmanagerConsumer(
+            charm=self,
+            relation_name="alertmanager",
+        )
+
+        self._register_event_handlers()
+
+    @property
+    def _scrape_job(self):
+        return [{"static_configs": [{"targets": [f"*:{self._port}"]}]}]
+
+    @property
+    def _prometheus_layer(self) -> Layer:
+        """Construct the pebble layer.
+
+        Returns:
+            a Pebble layer specification for the Prometheus workload container.
+        """
+        logger.debug("Building pebble layer")
+        layer_config = {
+            "summary": "Prometheus layer",
+            "description": "Pebble layer configuration for Prometheus",
+            "services": {
+                self._name: {
+                    "override": "replace",
+                    "summary": "prometheus daemon",
+                    "command": self._command(),
+                    "startup": "enabled",
+                }
+            },
+        }
+
+        return Layer(layer_config)
+
+    @property
+    def _external_url(self) -> str:
+        """Return the external hostname to be passed to ingress via the relation.
+
+        If we do not have an ingress, then use the pod ip as hostname.
+        The reason to prefer this over the pod name (which is the actual
+        hostname visible from the pod) or a K8s service, is that those
+        are routable virtually exclusively inside the cluster (as they rely)
+        on the cluster's DNS service, while the ip address is _sometimes_
+        routable from the outside, e.g., when deploying on MicroK8s on Linux.
+        """
+        if web_external_url := self.model.config.get("web_external_url"):
+            return web_external_url
+        if ingress_url := self.ingress.url:
+            return ingress_url
+        return f"http://{socket.getfqdn()}:{self._port}"
+
+    def _patch_kubernetes(self):
+        self.service_patch = KubernetesServicePatch(self, [(f"{self.app.name}", self._port)])
+        self.resources_patch = KubernetesComputeResourcesPatch(
```

Comment:
According to [some](https://stackoverflow.com/questions/38377276/instance-variables-in-methods-outside-the-constructor-python-why-and-how/38378757#38378757), it is best practice to initialize all fields inside the constructor itself.

Reply:
Ugh, Python really does promote some quite terrible practices 😅

Reply:
I really don't understand why Python promotes writing so lengthy init functions, but alright, rolling this back. It really could use more of the "do I really need to see the details or is it good enough to just get a summary?" extraction principles.